vnext/Microsoft.ReactNative/IReactNotificationService.h

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "ReactNotificationServiceHelper.g.h"
+#include <functional/functorRef.h>
+#include <object/objectRefCount.h>
+#include <winrt/Microsoft.ReactNative.h>
+#include <atomic>
+#include <map>
+#include <mutex>
+#include <vector>
+
+namespace winrt::Microsoft::ReactNative::implementation {
+
+struct ReactNotificationArgs : implements<ReactNotificationArgs, IReactNotificationArgs> {
+  ReactNotificationArgs(IReactNotificationSubscription const &subscription, IInspectable const &data) noexcept
+      : m_subscription{subscription}, m_data{data} {}
+
+  IReactNotificationSubscription Subscription() const noexcept {
+    return m_subscription;
+  }
+
+  IInspectable Data() const noexcept {
+    return m_data;
+  }
+
+ private:
+  IReactNotificationSubscription m_subscription;
+  IInspectable m_data;
+};
+
+struct ReactNotificationService : implements<ReactNotificationService, IReactNotificationService> {
+  ReactNotificationService();
+  explicit ReactNotificationService(IReactNotificationService const parentNotificationService) noexcept;
+  ~ReactNotificationService() noexcept;
+
+  IReactNotificationSubscription Subscribe(
+      IReactPropertyName const &notificationName,
+      IReactDispatcher const &dispatcher,
+      ReactNotificationHandler const &handler) noexcept;
+
+  void Unsubscribe(IReactNotificationSubscription const &subscription) noexcept;
+
+  void UnsubscribeAll() noexcept;
+
+  void SendNotification(
+      IReactPropertyName const &notificationName,
+      IInspectable const &sender,
+      IInspectable const &data) noexcept;
+
+ private:
+  // We treat subscription snapshots as immutable data.
+  using SubscriptionSnapshot = std::vector<IReactNotificationSubscription>;
+  using SubscriptionSnapshotPtr = Mso::RefCountedPtr<SubscriptionSnapshot>;
+
+  void ModifySubscriptions(
+      IReactPropertyName const &notificationName,
+      Mso::FunctorRef<SubscriptionSnapshot(SubscriptionSnapshot const &)> const &modifySnapshot);
+
+ private:
+  std::mutex m_mutex;
+  std::map<IReactPropertyName, SubscriptionSnapshotPtr> m_subscriptions;
+  IReactNotificationService m_parentNotificationService;
+};
+
+struct ReactNotificationServiceHelper {
+  ReactNotificationServiceHelper() = default;
+
+  static IReactNotificationService CreateNotificationService() noexcept;
+};
+
+struct ReactNotificationSubscription : implements<ReactNotificationSubscription, IReactNotificationSubscription> {
+  ReactNotificationSubscription(
+      weak_ref<ReactNotificationService> &&notificationService,
+      IReactPropertyName const &notificationName,
+      IReactDispatcher const &dispatcher,
+      ReactNotificationHandler const &handler) noexcept;
+  ~ReactNotificationSubscription() noexcept;
+
+  IReactPropertyName NotificationName() const noexcept;
+  IReactDispatcher Dispatcher() const noexcept;
+  bool IsSubscribed() const noexcept;
```

Comment:
>IsSubscribed [](start = 7, length = 12)

Why is this needed? Wouldn't it create race condition if someone calls if (IsSubscribed()) call()? (but someone unsubscribes in-between)

Reply:
Another way of putting it, what information does return value from this function give? If after returning false, there may still be callbacks en-route.

---
In reply to: [427621090](https://github.com/microsoft/react-native-windows/pull/4953#discussion_r427621090) [](ancestors = 427621090)

Reply:
It is useful inside of the notification handler. We check this property before we call the notification handler, but like you said the notification could be in-route. The only thread-safe to ensure that no code executed after unsubscription is to do another check inside of handler. It should be done only for special cases involved multi-threading, but at least we allow it.
In normal cases developers do not need to check this property.

---
In reply to: [427632411](https://github.com/microsoft/react-native-windows/pull/4953#discussion_r427632411) [](ancestors = 427632411,427621090)